README.md
[![Build Status](https://travis-ci.org/adnanyaqoobvirk/datadog-scala.png?branch=master)](https://travis-ci.org/adnanyaqoobvirk/datadog-scala)

# Datadog-Scala

A Scala library for interacting with the Datadog API using akka-http.

As of October 2014 this library covers all the methods in the [Datadog API Documentation](http://docs.datadoghq.com/api/).

# Example

```scala
import org.yaqoob.datadog.Client
import org.yaqoob.datadog.HttpAdapter

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer

implicit val defaultActorSystem = ActorSystem()
implicit val defaultMaterializer = ActorMaterializer()
implicit val executionContext = defaultActorSystem.dispatcher
    
val adapter = new HttpAdapter()
val client = new Client(apiKey = "XXX", appKey = "XXX", httpAdapter = adapter)
client.getAllTimeboards.foreach({ response =>
    println(response.body)
})
```


CHANGELOG.md
# 1.3.0
- Shifted the backend to akka-http

# 1.2.0
- Fixed error in README re: dependency, thanks [danbills](https://github.com/danbills)
- Upgrade to json4s 3.5.0, this may cause some compat woes hence the version bum, thanks [dsiegmann-ssc](https://github.com/dsiegmann-ssc)

# 1.1.3

- Release fuckery.

# 1.1.0

Bugfixes:
- Actually shutdown the actor system on `shutdown` even it wasn't supplied via [#6](https://github.com/gphat/datadog-scala/pull/6), thanks [ddelatre](https://github.com/ddelautre)
- Fix misspelling in README via [#7](https://github.com/gphat/datadog-scala/pull/7), thanks [earldouglas](https://github.com/earldouglas)
- Fix missing imports in README vi [*8](https://github.com/gphat/datadog-scala/pull/8), thanks [earldouglas](https://github.com/earldouglas)

Features:
- Add `addServiceCheck` method
- Add `deleteEvent` method
- Add `query` method


src/test/scala/TagSpec.scala
package test

import akka.actor.ActorSystem
import akka.http.scaladsl.model.HttpMethods
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.ActorMaterializer
import org.json4s._
import org.json4s.native.JsonMethods._
import org.specs2.mutable.Specification
import org.yaqoob.datadog.Client

import scala.concurrent.duration._
import scala.concurrent.Await

class TagSpec extends Specification {

  implicit val formats = DefaultFormats

  // Sequential because it's less work to share the client instance
  sequential

  "Client" should {

    implicit val defaultActorSystem = ActorSystem()
    implicit val defaultMaterializer = ActorMaterializer()
    implicit val executionContext = defaultActorSystem.dispatcher

    val adapter = new OkHttpAdapter()
    val client = new Client(
      apiKey = "apiKey",
      appKey = "appKey",
      httpAdapter = adapter
    )

    "handle get all tags" in {
      val res = Await.result(client.getAllTags, Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/tags/hosts?api_key=apiKey&application_key=appKey")
      adapter.getRequest must beSome.which(_.method == HttpMethods.GET)
    }

    "handle add tags for host" in {
      val res = Await.result(client.addTags("12345", Seq("foo:bar", "butt")), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/tags/hosts/12345?api_key=apiKey&application_key=appKey")

      val entity = Await.result(Unmarshal(adapter.getRequest.get.entity).to[String], Duration(1, "second"))
      val body = parse(entity)
      (body \ "tags")(0).extract[String] must beEqualTo("foo:bar")
      (body \ "tags")(1).extract[String] must beEqualTo("butt")

      adapter.getRequest must beSome.which(_.method == HttpMethods.POST)
    }

    "handle get tags for host" in {
      val res = Await.result(client.getTags("12345"), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/tags/hosts/12345?api_key=apiKey&application_key=appKey")
      adapter.getRequest must beSome.which(_.method == HttpMethods.GET)
    }

    "handle delete tags" in {
      val res = Await.result(client.deleteTags("12345"), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/tags/hosts/12345?api_key=apiKey&application_key=appKey")

      adapter.getRequest must beSome.which(_.method == HttpMethods.DELETE)
    }
  }
}

src/test/scala/UserSpec.scala
package test

import akka.actor.ActorSystem
import akka.http.scaladsl.model.HttpMethods
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.ActorMaterializer
import org.json4s._
import org.json4s.native.JsonMethods._
import org.specs2.mutable.Specification
import org.yaqoob.datadog.Client

import scala.concurrent.duration._
import scala.concurrent.Await

class UserSpec extends Specification {

  implicit val formats = DefaultFormats

  // Sequential because it's less work to share the client instance
  sequential

  "Client" should {

    implicit val defaultActorSystem = ActorSystem()
    implicit val defaultMaterializer = ActorMaterializer()
    implicit val executionContext = defaultActorSystem.dispatcher

    val adapter = new OkHttpAdapter()
    val client = new Client(
      apiKey = "apiKey",
      appKey = "appKey",
      httpAdapter = adapter
    )

    "handle invite users" in {
      val res = Await.result(client.inviteUsers(Seq("friend@example.com")), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/invite_users?api_key=apiKey&application_key=appKey")

      val entity = Await.result(Unmarshal(adapter.getRequest.get.entity).to[String], Duration(1, "second"))
      val body = parse(entity)
      (body \ "emails")(0).extract[String] must beEqualTo("friend@example.com")

      adapter.getRequest must beSome.which(_.method == HttpMethods.POST)
    }
  }
}

src/test/scala/AlertSpec.scala
package test

import akka.actor.ActorSystem
import org.yaqoob.datadog._
import akka.http.scaladsl.model.HttpMethods
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.ActorMaterializer
import org.json4s.DefaultFormats
import org.specs2.mutable.Specification
import org.yaqoob.datadog.Client

import scala.concurrent.duration._
import scala.concurrent.Await

class AlertSpec extends Specification {

  implicit val formats = DefaultFormats

  // Sequential because it's less work to share the client instance
  sequential

  "Client" should {

    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val executionContext = system.dispatcher

    val adapter = new OkHttpAdapter()
    val client = new Client(
      apiKey = "apiKey",
      appKey = "appKey",
      httpAdapter = adapter
    )

    "handle get all alerts" in {
      val res = Await.result(client.getAllAlerts, Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/alert?api_key=apiKey&application_key=appKey")
      adapter.getRequest must beSome.which(_.method == HttpMethods.GET)
    }

    "handle add alert" in {
      val res = Await.result(client.addAlert("POOP"), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/alert?api_key=apiKey&application_key=appKey")
      Await.result(Unmarshal(adapter.getRequest.get.entity).to[String], Duration(1, "second")) shouldEqual "POOP"

      adapter.getRequest must beSome.which(_.method == HttpMethods.POST)
    }

    "handle get alert" in {
      val res = Await.result(client.getAlert(12345), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/alert/12345?api_key=apiKey&application_key=appKey")
      adapter.getRequest must beSome.which(_.method == HttpMethods.GET)
    }

    "handle delete alert" in {
      val res = Await.result(client.deleteAlert(12345), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/alert/12345?api_key=apiKey&application_key=appKey")

      adapter.getRequest must beSome.which(_.method == HttpMethods.DELETE)
    }

    "handle update screenboard" in {
      val res = Await.result(client.updateAlert(12345, "POOP"), Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/alert/12345?api_key=apiKey&application_key=appKey")
      Await.result(Unmarshal(adapter.getRequest.get.entity).to[String], Duration(1, "second")) shouldEqual "POOP"

      adapter.getRequest must beSome.which(_.method == HttpMethods.PUT)
    }

    "handle mute all alerts" in {
      val res = Await.result(client.muteAllAlerts, Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/mute_alerts?api_key=apiKey&application_key=appKey")

      adapter.getRequest must beSome.which(_.method == HttpMethods.POST)
    }

    "handle unmute all alerts" in {
      val res = Await.result(client.unmuteAllAlerts, Duration(5, "second"))

      res.statusCode must beEqualTo(200)
      adapter.getRequest must beSome.which(_.uri.toString == "https://app.datadoghq.com/api/v1/unmute_alerts?api_key=apiKey&application_key=appKey")

      adapter.getRequest must beSome.which(_.method == HttpMethods.POST)
    }
  }
}

project/Publish.scala
import sbt._
import sbt.Keys._

object Publish {
  lazy val settings = Seq(
    homepage := Some(url("http://github.com/adnanyaqoobvirk/datadog-scala")),
    publishMavenStyle := true,
    publishTo := {
        val nexus = "https://oss.sonatype.org/"
        if (isSnapshot.value)
            Some("snapshots" at nexus + "content/repositories/snapshots")
        else
            Some("releases"  at nexus + "service/local/staging/deploy/maven2")
    },
    publishArtifact in Test := false,
    pomIncludeRepository := Function.const(false),
    pomExtra := (
      <licenses>
       <license>
        <name>MIT</name>
        <url>https://opensource.org/licenses/MIT</url>
        <distribution>repo</distribution>
       </license>
      </licenses>
      <scm>
        <url>git@github.com:adnanyaqoobvirk/datadog-scala.git</url>
        <connection>scm:git:git@github.com:adnanyaqoobvirk/datadog-scala.git</connection>
      </scm>
      <developers>
        <developer>
          <name>Cory Watson</name>
          <email>github@onemogin.com</email>
          <organization>Cory Industries Ltd Inc</organization>
          <organizationUrl>https://onemogin.com</organizationUrl>
        </developer>
        <developer>
          <name>Adnan Yaqoob</name>
          <email>adnan@yaqoob.org</email>
          <organization>Yaqoob.org</organization>
          <organizationUrl>https://yaqoob.org</organizationUrl>
        </developer>
      </developers>
    )
  )
}


src/main/scala/org/yaqoob/datadog/HttpAdapter.scala
package org.yaqoob.datadog

import java.util.concurrent.TimeUnit

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.client.RequestBuilding
import akka.http.scaladsl.model._
import akka.http.scaladsl.model.Uri.{Authority, Host, Path, Query}
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Sink, Source}
import akka.util.Timeout

import org.yaqoob.datadog.common.Response

import scala.concurrent.{ExecutionContextExecutor, Future}

/**
  * Created by Adnan Yaqoob on 13/04/2017.
  */
class HttpAdapter(
                   httpTimeoutSeconds: Int = 10
                 )(implicit system: ActorSystem, materializer: ActorMaterializer, executtionContext: ExecutionContextExecutor) {

  // Akka's Ask pattern requires an implicit timeout to know
  // how long to wait for a response.
  implicit val timeout = Timeout(httpTimeoutSeconds, TimeUnit.SECONDS)

  val httpPool = Http().superPool[Unit]()

  def doRequest(
                 scheme: String,
                 authority: String,
                 path: String,
                 method: String,
                 body: Option[String] = None,
                 params: Map[String,Option[String]] = Map.empty,
                 contentType: String = "json"): Future[Response] = {

    // Turn a map of string,opt[string] into a map of string,string which is
    // what Query wants
    val filteredParams = params.filter(
      // Filter out keys that are None
      _._2.isDefined
    ).map(
      // Convert the remaining tuples to str,str
      param => param._1 -> param._2.get
    )
    // Make a Uri
    val finalUrl = Uri(
      scheme = scheme,
      authority = Authority(host = Host(authority)),
      path = Path("/api/v1/" + path)
    ).withQuery(Query(filteredParams))

    // Use the provided case classes from spray-client
    // to construct an HTTP request of the type needed.
    val httpRequest: HttpRequest = method match {
      case "DELETE" => RequestBuilding.Delete(finalUrl, body)
      case "GET" => RequestBuilding.Get(finalUrl, body)
      case "POST" => contentType match {
        case "json" => RequestBuilding.Post(finalUrl, body.map({
          b => HttpEntity(ContentTypes.`application/json`, b)
        }).getOrElse(HttpEntity.Empty))
        case _ => {
          // This is going to be a form-encoded post. There's only one
          // API call that works this way (ugh) so I'm not going to worry
          // too much about making this work as cleanly as the rest of the
          // stuff. (IMO)
          val formUrl = Uri(
            scheme = scheme,
            authority = Authority(host = Host(authority)),
            path = Path("/api/v1/" + path)
          )

          RequestBuilding.Post(formUrl, FormData(filteredParams))
        }
      }
      case "PUT" => RequestBuilding.Put(finalUrl, HttpEntity(ContentTypes.`application/json`, body.get))
      case _ => throw new IllegalArgumentException("Unknown HTTP method: " + method)
    }

    system.log.debug("%s: %s".format(method, finalUrl))
    // For spelunkers, the ? is a function of the Akka "ask pattern". Unlike !
    // it waits for a response in the form of a future. In this case we're
    // sending along a case class representing the type of HTTP request we want
    // to do and something down in the guts of the actors handles it and gets
    // us a response.
    doHttp(httpRequest)
  }

  def doHttp(request: HttpRequest): Future[Response] = {
    Source.single((request, ()))
      .via(httpPool)
      .map { case (response, _) => response.get }
      .runWith(Sink.head)
      .flatMap(
        res =>
          Unmarshal(res.entity).to[String].map(
            entity =>
              Response(
                statusCode = res.status.intValue,
                entity
              )
          )
      )
  }
}


src/main/scala/org/yaqoob/datadog/Client.scala
package org.yaqoob.datadog

import org.yaqoob.datadog.common.{Metric, Response}
import org.json4s.FieldSerializer._
import org.json4s.JsonDSL._
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization
import org.json4s.jackson.Serialization.write


import scala.concurrent.Future

/**
  * Created by Adnan Yaqoob on 13/04/2017.
  */
class Client(
              scheme: String = "https",
              authority: String = "app.datadoghq.com",
              apiKey: String,
              appKey: String,
              httpAdapter: HttpAdapter
            ){

  val metricSerializer = FieldSerializer[Metric](
    renameTo("metricType", "metric_type"),
    renameFrom("metric_type", "metricType")
  )

  implicit val formats = Serialization.formats(NoTypeHints) + metricSerializer

  def addComment(
                  message: String, handle: Option[String] = None,
                  relatedEventId: Option[Long] = None
                ): Future[Response] = {

    val json =
      ("message" -> message) ~
        ("handle" -> handle) ~
        ("related_event_id" -> relatedEventId)

    val path = Seq("comments").mkString("/")
    doRequest(path = path, method = "POST", body = Some(write(json)))
  }

  def addEvent(
                title: String, text: String, dateHappened: Option[String] = None,
                priority: Option[String] = None, tags: Option[Seq[String]] = None,
                alertType: Option[String] = None, aggregationKey: Option[String] = None,
                sourceTypeName: Option[String] = None
              ): Future[Response] = {

    val json =
      ("title" -> title) ~
        ("text" -> text) ~
        ("date_happened" -> dateHappened) ~
        ("priority" -> priority) ~
        ("tags" -> tags) ~
        ("alert_type" -> alertType) ~
        ("aggregation_key" -> aggregationKey) ~
        ("source_type_name" -> sourceTypeName)

    val path = Seq("events").mkString("/")
    doRequest(path = path, method = "POST", body = Some(compact(render(json))))
  }

  def addMetrics(series: Seq[Metric]): Future[Response] = {

    val json =
      "series" -> series.map { s =>
        ("metric" -> s.name) ~
          ("points" -> s.points.map(tuple => Seq(JInt(tuple._1), JDouble(tuple._2)))) ~
          ("metric_type" -> s.metricType) ~
          ("tags" -> s.tags) ~
          ("host" -> s.host)
      }

    val path = Seq("series").mkString("/")
    doRequest(path = path, method = "POST", body = Some(write(json)))
  }

  def addServiceCheck(
                       check: String, hostName: String, status: Int,
                       timestamp: Option[Int] = None, message: Option[String] = None,
                       tags: Option[Seq[String]] = None
                     ): Future[Response] = {

    val path = Seq("check_run").mkString("/")
    doRequest(path = path, method = "POST", params = Map(
      "check" -> Some(check),
      "host_name" -> Some(hostName),
      "status" -> Some(status.toString),
      "timestamp" -> timestamp.map { _.toString },
      "message" -> message,
      "tags" -> tags.map({ tees => tees.mkString(",") })
    ))
  }

  def addScreenboard(board: String): Future[Response] = {

    val path = Seq("screen").mkString("/")
    doRequest(path = path, method = "POST", body = Some(board))
  }

  // XXX source
  def addTags(hostId: String, tags: Seq[String]): Future[Response] = {

    val json = "tags" -> tags

    val path = Seq("tags", "hosts", hostId).mkString("/")
    doRequest(path = path, method = "POST", body = Some(write(json)))
  }

  def addAlert(alert: String): Future[Response] = {

    val path = Seq("alert").mkString("/")
    doRequest(path = path, method = "POST", body = Some(alert))
  }

  def addTimeboard(board: String): Future[Response] = {

    val path = Seq("dash").mkString("/")
    doRequest(path = path, method = "POST", body = Some(board))
  }

  def deleteAlert(alertId: Long): Future[Response] = {

    val path = Seq("alert", alertId).mkString("/")
    doRequest(path = path, method = "DELETE")
  }

  def deleteComment(commentId: Long): Future[Response] = {

    val path = Seq("comments", commentId).mkString("/")
    doRequest(path = path, method = "DELETE")
  }

  def deleteEvent(eventId: Long): Future[Response] = {

    val path = Seq("events", eventId).mkString("/")
    doRequest(path = path, method = "DELETE")
  }

  def deleteScreenboard(boardId: Long): Future[Response] = {

    val path = Seq("screen", boardId).mkString("/")
    doRequest(path = path, method = "DELETE")
  }

  def deleteTags(hostId: String): Future[Response] = {

    val path = Seq("tags", "hosts", hostId).mkString("/")
    doRequest(path = path, method = "DELETE")
  }

  def deleteTimeboard(boardId: Long): Future[Response] = {

    val path = Seq("dash", boardId).mkString("/")
    doRequest(path = path, method = "DELETE")
  }

  def getAlert(id: Long): Future[Response] = {
    val path = Seq("alert", id.toString).mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getScreenboard(id: Long): Future[Response] = {
    val path = Seq("screen", id.toString).mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getTimeboard(id: Long): Future[Response] = {
    val path = Seq("dash", id.toString).mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getAllAlerts(): Future[Response] = {
    val path = Seq("alert").mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getAllScreenboards(): Future[Response] = {
    val path = Seq("screen").mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getAllTags: Future[Response] = {
    val path = Seq("tags", "hosts").mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getAllTimeboards(): Future[Response] = {
    val path = Seq("dash").mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getEvent(eventId: Long): Future[Response] = {
    val path = Seq("events", eventId.toString).mkString("/")
    doRequest(path = path, method = "GET")
  }

  def getEvents(
                 start: Long, end: Long, priority: Option[String] = None,
                 sources: Option[Seq[String]] = None, tags: Option[Seq[String]] = None
               ): Future[Response] = {

    val params = Map(
      "start" -> Some(start.toString),
      "end" -> Some(end.toString),
      "priority" -> priority,
      "sources" -> sources.map({ esses => esses.mkString(",") }),
      "tags" -> tags.map({ tees => tees.mkString(",") })
    )

    val path = Seq("events").mkString("/")
    doRequest(path = path, method = "GET", params = params, contentType = "form")
  }

  // XXX source and by_source?
  def getTags(hostId: String): Future[Response] = {
    val path = Seq("tags", "hosts", hostId).mkString("/")
    doRequest(path = path, method = "GET")
  }

  def graphSnapshot(query: String, start: Long, end: Long, eventQuery: Option[String]): Future[Response] = {
    val path = Seq("graph", "snapshot").mkString("/")
    doRequest(path = path, method = "GET", params = Map(
      "metric_query" -> Some(query),
      "start" -> Some(start.toString),
      "end" -> Some(end.toString),
      "event_query" -> eventQuery
    ))
  }

  def inviteUsers(emails: Seq[String]): Future[Response] = {

    val json = ("emails" -> emails)

    val path = Seq("invite_users").mkString("/")
    doRequest(path = path, method = "POST", body = Some(write(json)))
  }

  def muteAllAlerts: Future[Response] = {

    val path = Seq("mute_alerts").mkString("/")
    doRequest(path = path, method = "POST")
  }

  def query(query: String, from: Int, to: Int): Future[Response] = {
    val path = Seq("query")
    doRequest(path = "query", method = "GET", params = Map(
      "query" -> Some(query),
      "from" -> Some(from.toString),
      "to" -> Some(to.toString)
    ))
  }

  def search(query: String): Future[Response] = {
    val path = Seq("search").mkString("/")
    doRequest(path = path, method = "GET", params = Map("q" -> Some(query)))
  }

  def unmuteAllAlerts: Future[Response] = {

    val path = Seq("unmute_alerts").mkString("/")
    doRequest(path = path, method = "POST")
  }

  def updateAlert(alertId: Long, alert: String): Future[Response] = {

    val path = Seq("alert", alertId).mkString("/")
    doRequest(path = path, method = "PUT", body = Some(alert))
  }

  def updateComment(
                     commentId: Long, message: Option[String], handle: Option[String] = None,
                     relatedEventId: Option[Long] = None
                   ): Future[Response] = {

    val json =
      ("message" -> message) ~
        ("handle" -> handle)

    val path = Seq("comments", commentId).mkString("/")
    doRequest(path = path, method = "PUT", body = Some(write(json)))
  }

  def updateScreenboard(boardId: Long, board: String): Future[Response] = {

    val path = Seq("screen", boardId).mkString("/")
    doRequest(path = path, method = "PUT", body = Some(board))
  }

  def updateTimeboard(boardId: Long, board: String): Future[Response] = {

    val path = Seq("dash", boardId).mkString("/")
    doRequest(path = path, method = "PUT", body = Some(board))
  }

  private def doRequest(
                         path: String,
                         method: String,
                         body: Option[String] = None,
                         params: Map[String,Option[String]] = Map.empty,
                         contentType: String = "json") = {

    httpAdapter.doRequest(
      method = method, scheme = scheme, authority = authority, path = path,
      body = body, params = params ++ Map("api_key" -> Some(apiKey), "application_key" -> Some(appKey)),
      contentType = contentType
    )
  }
}

